server/channels/app/response_transfer.go

```diff
@@ -25,8 +33,32 @@ func (rt *PluginResponseWriter) Header() http.Header {
 	return rt.headers
 }
 
+func (rt *PluginResponseWriter) Write(data []byte) (int, error) {
+	// Signal that headers are ready on first write if not already done
+	select {
+	case <-rt.HeadersReady:
+		// Already closed
+	default:
+		close(rt.HeadersReady)
+	}
+	return rt.pipeWriter.Write(data)
+}
+
 func (rt *PluginResponseWriter) WriteHeader(statusCode int) {
-	rt.statusCode = statusCode
+	if rt.statusCode == 0 {
+		rt.statusCode = statusCode
+		close(rt.HeadersReady)
```

Comment:
Unlikely to call `WriteHeader` more than once, and unlikely to do so concurrently, but should we guard against the resulting panic by using the pattern above instead of closing conditionally only on `rt.statusCode`?